core/src/main/java/org/apache/calcite/sql/fun/SqlJsonValueFunction.java

```diff
@@ -0,0 +1,196 @@
+    return SqlOperandCountRanges.between(5, 6);
+  }
+
+  @Override public boolean checkOperandTypes(SqlCallBinding callBinding,
+                                             boolean throwOnFailure) {
+    RelDataType defaultValueOnEmptyType =
+        callBinding.getValidator().getValidatedNodeType(callBinding.operand(2));
+    RelDataType defaultValueOnErrorType =
+        callBinding.getValidator().getValidatedNodeType(callBinding.operand(4));
+    RelDataType returnType = callBinding.getValidator().deriveType(
+        callBinding.getScope(), callBinding.operand(5));
+    if (canCastFrom(callBinding, throwOnFailure, defaultValueOnEmptyType, returnType)) {
+      return false;
+    }
+    if (canCastFrom(callBinding, throwOnFailure, defaultValueOnErrorType, returnType)) {
+      return false;
+    }
+    return true;
+  }
+
+  @Override public RelDataType inferReturnType(SqlOperatorBinding opBinding) {
+    assert opBinding.getOperandCount() == 5 || opBinding.getOperandCount() == 6;
+    RelDataType ret;
+    if (opBinding.getOperandCount() == 6) {
+      ret = opBinding.getOperandType(5);
+    } else {
+      ret = opBinding.getTypeFactory().createSqlType(SqlTypeName.ANY);
+    }
+    return opBinding.getTypeFactory().createTypeWithNullability(
+        ret,
+        true);
+  }
+
+  @Override public String getSignatureTemplate(int operandsCount) {
+    assert operandsCount == 5 || operandsCount == 6;
+    if (operandsCount == 6) {
+      return "{0}({1} RETURNING {6} {2} {3} ON EMPTY {4} {5} ON ERROR)";
+    }
+    return "{0}({1} {2} {3} ON EMPTY {4} {5} ON ERROR)";
+  }
+
+  @Override public void unparse(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    assert call.operandCount() == 5 || call.operandCount() == 6;
+    final SqlWriter.Frame frame = writer.startFunCall(getName());
+    call.operand(0).unparse(writer, 0, 0);
+    if (!returnAny) {
+      writer.keyword("RETURNING");
+      call.operand(5).unparse(writer, 0, 0);
+    }
+    unparseEnum(writer, call.operand(1));
+    if (isDefaultLiteral(call.operand(1))) {
+      call.operand(2).unparse(writer, 0, 0);
+    }
+    writer.keyword("ON");
+    writer.keyword("EMPTY");
+    unparseEnum(writer, call.operand(3));
+    if (isDefaultLiteral(call.operand(3))) {
+      call.operand(4).unparse(writer, 0, 0);
+    }
+    writer.keyword("ON");
+    writer.keyword("ERROR");
+    writer.endFunCall(frame);
+  }
+
+  private void unparseEnum(SqlWriter writer, SqlLiteral literal) {
+    writer.keyword(((Enum) literal.getValue()).name());
+  }
+
+  private boolean isDefaultLiteral(SqlLiteral literal) {
+    return literal.getValueAs(SqlJsonValueEmptyOrErrorBehavior.class)
+        == SqlJsonValueEmptyOrErrorBehavior.DEFAULT;
+  }
+
+  private boolean canCastFrom(SqlCallBinding callBinding, boolean throwOnFailure,
+                              RelDataType inType, RelDataType outType) {
+    if (!SqlTypeUtil.canCastFrom(outType, inType, true)) {
```

Comment:
Please invert `if` condition and put shorter branches to the top in the cases like that.

The following is better than heavily nested code:
```
if (canCastFrom){
  return true;
}
if (!throwOnFailure) {
  return false;
}
throw ...
```

Reply:
+1 thanks, and found a bug in line 120 and line 123 in the same file: the condition in if clause should be negative: ` if(canCastFrom) -> if (!canCastFrom)`

I will fix and add unit test for that.